### ISSUE TITLE
fparser.api.parse fails for long comment blocks

### ISSUE THREAD

**Issue by TeranIvy** [stfc/fparser#280]
When parsing a kernel that has a comment header longer than 161 lines PSyclone reports the incorrect Fortran error, e.g.
```
"Parse Error: Failed to parse kernel code 'test_fparser_kernel_mod.f90'. Is the Fortran correct?"
```
The error is raised in the [`psyclone.parse.kernel.get_kernel_parse_tree`](https://github.com/stfc/PSyclone/blob/master/src/psyclone/parse/kernel.py#L146) function. However, this is actually an fparser failure but that is not visible from PSyclone as the fparser logging is [disabled](https://github.com/stfc/PSyclone/blob/master/src/psyclone/parse/kernel.py#L160).

`psyclone.parse.kernel.get_kernel_parse_tree` calls [`fparser.api.parse`](https://github.com/stfc/fparser/blob/master/src/fparser/api.py#L150) function. This function, in turn, calls the [`fparser.one.parsefortran.FortranParser.parse`](https://github.com/stfc/fparser/blob/master/src/fparser/one/parsefortran.py#L123) function after creating a [reader object](https://github.com/stfc/fparser/blob/master/src/fparser/api.py#L83) from a source file.

Attached is the test example [test_fparser.zip](https://github.com/stfc/fparser/files/5771679/test_fparser.zip) with the Python test script `test_parsekernel.py` that illustrates the fparser error and uses the attached test kernel `test_fparser_kernel_mod.f90` to trigger the error.

The attached script tries to parse the kernel by two methods:

1. Directly calling the `fparser.api.parse` function as it is done in [`psyclone.parse.kernel.get_kernel_parse_tree`](https://github.com/stfc/PSyclone/blob/master/src/psyclone/parse/kernel.py#L162);
2. Executing step-by-step commands from the `fparser.api.parse` function that calls [`fparser.api.get_reader`](https://github.com/stfc/fparser/blob/master/src/fparser/api.py#L185) and then [`fparser.one.parsefortran.FortranParser.parse`](https://github.com/stfc/fparser/blob/master/src/fparser/api.py#L189) function.

The outcome of these methods depends on the value of `ignore_comments` parameter in the [`fparser.api.get_reader`](https://github.com/stfc/fparser/blob/master/src/fparser/api.py#L84) and [`fparser.api.parse`](https://github.com/stfc/fparser/blob/master/src/fparser/api.py#L151) functions in an fparser installation. Currently both instances are set to `True` in the `fparser.api`script (and hence in an fparser installation) so a kernel parse tree does not include comments. 

- With the default `ignore_comments=True` setting in `get_reader` and `parse` functions the first method, a direct call to `fparser.api.parse` function fails with 

  ```
  File "test_parsekernel.py", line 50, in <module>
    parse_tree = fpapi.parse(filepath)
  File "/data/users/lfric/modules/packages.rhel7/psyclone/1.9.0/lib/python3.6/site-packages/fparser/api.py", line 189, in parse
    parser.parse()
  File "/data/users/lfric/modules/packages.rhel7/psyclone/1.9.0/lib/python3.6/site-packages/fparser/one/parsefortran.py", line 143, in parse
    raise error
  File "/data/users/lfric/modules/packages.rhel7/psyclone/1.9.0/lib/python3.6/site-packages/fparser/one/parsefortran.py", line 128, in parse
    self.block = BeginSource(self)
  File "/data/users/lfric/modules/packages.rhel7/psyclone/1.9.0/lib/python3.6/site-packages/fparser/common/base_classes.py", line 776, in __init__
    Statement.__init__(self, parent, item)
  File "/data/users/lfric/modules/packages.rhel7/psyclone/1.9.0/lib/python3.6/site-packages/fparser/common/base_classes.py", line 605, in __init__
    self.process_item()
  File "/data/users/lfric/modules/packages.rhel7/psyclone/1.9.0/lib/python3.6/site-packages/fparser/one/block_statements.py", line 324, in process_item
    self.fill(end_flag=True)
  File "/data/users/lfric/modules/packages.rhel7/psyclone/1.9.0/lib/python3.6/site-packages/fparser/common/base_classes.py", line 827, in fill
    if self.process_subitem(item):
  File "/data/users/lfric/modules/packages.rhel7/psyclone/1.9.0/lib/python3.6/site-packages/fparser/one/block_statements.py", line 366, in process_subitem
    return BeginStatement.process_subitem(self, item)
  File "/data/users/lfric/modules/packages.rhel7/psyclone/1.9.0/lib/python3.6/site-packages/fparser/common/base_classes.py", line 928, in process_subitem
    self.handle_unknown_item(item)
  File "/data/users/lfric/modules/packages.rhel7/psyclone/1.9.0/lib/python3.6/site-packages/fparser/common/base_classes.py", line 945, in handle_unknown_item
    raise AnalyzeError(message)
  fparser.common.utils.AnalyzeError: While processing 'test_fparser_kernel_mod.f90' (mode='free')..
  219:  end subroutine test_fparser_code
  220:
  221:end module test_fparser_kernel_mod <== no parse pattern found for "end module test_fparser_kernel_mod" in 'BeginSource' block.

  ```

  The second method with the step-by-step commands to create reader and parser objects from `fparser.api.parse` works fine.

- When an fparser installation is hacked to set `ignore_comments` to `False`  in `fparser.api.get_reader` and `fparser.api.parse` functions , the first method, a direct call to `fparser.api.parse` function works fine. The second method with the step-by-step commands to create instances of reader and parser objects also works fine if  `ignore_comments = False` in the attached script. If it is `True`, however, it fails with the same error as the first method with the default `ignore_comments=False`  setting in the `fparser.api` installed module.

When the [maximum depth of the Python interpreter stack](https://docs.python.org/3/library/sys.html#sys.setrecursionlimit) is increased from the default limit of 1000 both the direct call to `fparser.api.parse` and step-by-step approach work regardless of the `ignore_comments` setting in fparser installation and the test script (demonstrated by enbling the commented line `sys.setrecursionlimit(1100)`).

Parsing the offending kernel by fparser2 (`fparser2 test_fparser_kernel_mod.f90`) works fine.

**Comment by rupertford** [comment]
I've had an initial look at this and there seems to be an issue with the reader when there are many consecutive comment lines and the readers `ignore_comments` argument is set to `True`. If the number is too large the reader will raise an exception.

As @TeranIvy noticed this can be resolved by increasing the recursion limit to a high enough number. However, the more consecutive comments there are, the larger this value needs to be set to, so it should only be considered a temporary fix.

When looking more closely at fparser1 (which currently fails), the parser itself actually has its own `ignore_comments` flag which does not suffer from this bug. Therefore a potential temporary solution would be to always set the `ignore_comments` flag for the `reader` to False and control whether comments are output using the `ignore_comments` flag for `fparser1`. Something like ...

```
reader = FortranFileReader("tmp.f90", ignore_comments=False)
parser = parsefortran.FortranParser(reader, ignore_comments=XXX)
```

Given that it is an issue with the reader, the obvious question is why it does not fail with fparser2. Well the answer is that it does, but fparser2 somehow manages to ignore the error and continue.

What we see in the fparser2 output is something like ...

```
CRITICAL:fparser.common.readfortran:While processing 'tmp.f90' (mode='free')..
193:  !> @param[in]     undf_adspc2          Number of unique DOFs for any_d_2 space
194:  !> @param[in]     map_adspc2           Dofmap for the cell at the base of the column for any_d_2 space
195:  !> Lines 31-35 of a very long subroutine comment header
196:  !> @param[in]     nlayers              Number of layers <== while processing line
CRITICAL:fparser.common.readfortran:STOPPED READING 
```

Whilst it is good that fparser2 manages to continue somehow (I need to see what it is doing), fparser2 does not have its own `ignore_comments` option and we don't really want to have to suppress errors, even if the result is OK. Therefore I think I should try to sort out the problem with the reader in the first instance.

**Comment by rupertford** [comment]
Created branch [280_comments_bug](https://github.com/stfc/fparser/tree/280_comments_bug)

**Comment by rupertford** [comment]
The problem was that when a comment line was matched, it was added to the fifo buffer and then the next comment line was matched until all consecutive comment lines are matched. This leads to a lot of recursion. As far as I can tell there is no reason to do it this way as each line is treated independently (i.e. there are no multi-line comments).  Therefore I modified the code to return comments line by line. The PSyclone tests seem fine so hopefully the changes are OK.
